ignite/metrics/precision.py

```diff
@@ -26,16 +27,48 @@ def reset(self):
 
     def update(self, output):
         y_pred, y = output
-        num_classes = y_pred.size(1)
-        indices = torch.max(y_pred, 1)[1]
-        correct = torch.eq(indices, y)
-        pred_onehot = to_onehot(indices, num_classes)
-        all_positives = pred_onehot.sum(dim=0)
+        dtype = y_pred.type()
+
+        if not (y.ndimension() == y_pred.ndimension() or y.ndimension() + 1 == y_pred.ndimension()):
+            raise ValueError("y must have shape of (batch_size, ...) " +
+                             "and y_pred must have shape of (batch_size, num_classes, ...) or " +
+                             "(batch_size, ...).")
+
+        if y.ndimension() > 1 and y.shape[1] == 1:
+            y = y.squeeze(dim=1)
+
+        if y_pred.ndimension() > 1 and y_pred.shape[1] == 1:
+            y_pred = y_pred.squeeze(dim=1)
+
+        y_shape = y.shape
+        y_pred_shape = y_pred.shape
+
+        if y.ndimension() + 1 == y_pred.ndimension():
+            y_pred_shape = (y_pred_shape[0],) + y_pred_shape[2:]
+
+        if not (y_shape == y_pred_shape):
+            raise ValueError("y and y_pred must have compatible shapes.")
+
+        if y_pred.ndimension() == y.ndimension() + 1:
```

Comment:
Can't we use the same trick as in Accuracy here ?
I think it would be rather weird to have `threshold_function` for Precision and Recall and not for Accuracy...
IMO, we do not need `threshold_function` for these metrics. Correct me if I'm wrong.

Reply:
Incorporated! Changed to similar to accuracy